scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/EquationSolvingAlgorithms.java

```diff
@@ -64,79 +82,204 @@ public static class NoSolutionException extends Exception{
     public NoSolutionException() {
       super();
     }
-    public NoSolutionException(String errorMessage) {
-      super(errorMessage);
-    }
   }
 
-  public static class SecantDurationAlgorithm<History> implements SecantAlgorithm<Duration, History>{
-
-    public RootFindingResult<Duration, History> findRoot(
-        Function<Duration, History> f,
-        History history,
-        Duration x0,
-        Duration x1,
-        Duration y,
-        Duration toleranceYLow,
-        Duration toleranceYHigh,
-        Duration xLow,
-        Duration xHigh,
-        int maxNbIterations)
-    throws ZeroDerivativeException, InfiniteDerivativeException, DivergenceException, ExceededMaxIterationException, NoSolutionException
+  public static class SecantDurationAlgorithm<Metadata> implements SecantAlgorithm<Duration, Metadata>{
+
+    private static RandomGenerator randomGenerator = RandomGeneratorFactory.of("Random").create(956756789);
+
+    /**
+     * Randomly selects a value in the interval [bound1, bound2]
+     * @param bound1 the first bound
+     * @param bound2 the second bound
+     * @return a value chosen randomly
+     */
+    private Duration chooseRandomX(final Duration bound1, final Duration bound2){
+      var low = bound1;
+      var high = bound2;
+      if(low.isEqualTo(high)) return low;
+      if(bound1.longerThan(bound2)) { low = bound2; high = bound1; }
+      return Duration.of(
+          randomGenerator.nextLong(low.in(Duration.MICROSECONDS), high.in(Duration.MICROSECONDS)),
+          Duration.MICROSECONDS);
+    }
+
+    private record IteratingResult(FunctionCoordinate<Duration> result, int nbIterationsPerformed){}
+
+    /**
+     * Querying Function.valueAt may lead to a discontinuity. This procedure starts at an initial x value
+     * and stops only when the value returned is not a discontinuity or the maximum number of iterations has been reached
+     * Kind of an infaillible valueAt with a limited number of iterations
+     * @param function the function we are trying to call
+     * @param init the initial x value
+     * @param min the lower bound of the domain of x
+     * @param max the upper bound of the domain of x
+     * @param history the querying history of f
+     * @param maxIteration the maximum number of iteration possible
+     * @return a coordinate (x, f(x)) s.t. f is continuous at x.
+     * @throws ExceededMaxIterationException
+     */
+    private IteratingResult tryValueAtUntilExhaustion(
```

Comment:
Minor, can ignore: I'm still trying to think if there's a more concise name that could be used. Would `valueAround` (or maybe `getValueAround`/`findValueAround`) fit? The `Around` communicates that the method doesn't necessarily return the function's output at `init`.

Reply:
Not sure about that. I wanted to call it `infaillibleValueAt` but it can still fail if too many iterations...

Reply:
`nextValueAt` it is